log/logger.go

```diff
@@ -32,4 +32,10 @@ type (
 		Warn(msg string, keyvals ...interface{})
 		Error(msg string, keyvals ...interface{})
 	}
+
+	// WithSkipCallers is an optional interface that a Logger can implement that
+	// indicate the number of stack frames to skip.
+	WithSkipCallers interface {
```

Comment:
Some users brought up the logger they pass to temporal already, unintentionally, implements this interface and this would be a breaking change for them. I may have to rename this or add an extra guard method to the interface to ensure people mean to implement it.

Reply:
I don't think it's a breaking change, it's not like you added this on the existing interface. If they do implement this interface exactly, with return type and all, great this works because the behavior should be obvious. If they make have the same method name, it's harmless because it won't match the interface. And if they _did_ want skip support but already have a method with this name, they can wrap-and-embed, only providing this function which shadows their other one for this specific use.

Reply:
Hrmm, I wonder how they're getting their logger from the `ReplayLogger` to make that call. Regardless I think it's a bit rough for us not to be able to add an interface because someone else had that method and was using it already.

But if this is a big enough concern, I guess we can change to `WithCallerSkip` which maybe should be the name anyways since this should probably never mutate the existing logger and we need to document that too. Granted if that is a problem too because maybe they used that method name also, then what do we do if we can never add extensibility points because of this fear?

Reply:
>Granted if that is a problem too because maybe they used that method name also, then what do we do if we can never add extensibility points because of this fear?

If you pick method names that are used by other common loggers your asking for this kind of issue. It's an easy problem to work around, two simple solution are pick a method name users are very unlikely to have implemented themselves like `AddCallerSkipTemporal` or require the interface to have an addition method like `TemporalSkipLogger`

Reply:
That's a bit rough I think, Is `WithCallerSkip` unique enough? (also, can we doc that this method should return a new logger with the skip amount but not mutate the current one?)

Reply:
> That's a bit rough I think. Is WithCallerSkip unique/good enough? 

Since `WithCallerSkip` is used by zap I don't think it is unique.

> (also, can we doc that this method should return a new logger with the skip amount but not mutate the current one?)

We should  document that it may not mutate the existing one.

Reply:
>AddCallerSkip is there, but not WithCallerSkip

Ah I was looking at a wrapper for Zap, not zap itself.  Still a quicks search shows other loggers define `WithCallerSkip` with the same signature as us.

Reply:
But they won't match the interface then since the return type is different. The only concern I think is those that match the interface and get misused when _we_ call them internally. I'm just trying to avoid those that have written this exact method signature with the same return value and all.

Reply:
if their `WithCallerSkip` returns a `logger` that matches our interface then they would match the interface no?

Reply:
No it has to be the exact signature. Return types are not covariant when matching interfaces. E.g. https://go.dev/play/p/USVfNSHJUIk (ref https://go.dev/ref/spec#Interface_types on exact method set match). That's why I considered it mostly not a big deal above: "If they make have the same method name but different params/return, it's harmless because it won't match the interface".